docs/standard/microservices-architecture/multi-container-microservice-net-applications/implementing-api-gateways-with-ocelot.md

```diff
@@ -0,0 +1,566 @@
+
+Therefore, for many medium- and large-size applications, using a custom-built API Gateway product is usually a good approach, but not as a single monolithic aggregator or unique central custom API Gateway unless that API Gateway allows multiple independent configuration areas for the several development teams creating autonomous microservices.
+
+
+### Sample microservices/containers to reroute through the API Gateways
+
+As an example, 'eShopOnContainers' has around six internal microservice-types that have to be published through the API Gateways, as shown in the following image.
+ 
+![](./media/image29.png)
+
+**Figure 8-28.** Microservice folders in eShopOnContainers solution in Visual Studio
+
+About the Identity service, in the design it is left out of the API Gateway routing because it is the only cross-cutting concern in the system, but with Ocelot it is also possible to include it as part of the rerouting lists.
+
+All those services are currently implemented as ASP.NET Core Web API services, as you can tell because of the code. Let’s focus on one of the microservices like the Catalog microservice code.
+
+![](./media/image30.png)
+
+**Figure 8-29.** Sample Web API microservice (Catalog microservice)
+
+You can see that the Catalog microservice is a typical ASP.NET Core Web API project with several controllers and methods like in the following code.
+
+```csharp
+[HttpGet]
+[Route("items/{id:int}")]
+[ProducesResponseType((int)HttpStatusCode.NotFound)]
+[ProducesResponseType(typeof(CatalogItem),(int)HttpStatusCode.OK)]
+public async Task<IActionResult> GetItemById(int id)
+{
+    if (id <= 0)
+    {
+        return BadRequest();
+    }
+    var item = await _catalogContext.CatalogItems.
+                                          SingleOrDefaultAsync(ci => ci.Id == id);
+    //…
+
+    if (item != null)
+    {
+        return Ok(item);
+    }
+    return NotFound();
+}
+```
+The Http request will end up running that kind of C# code accessing the microservice database plus any additional action.
+
+In regards the microservice URL, when the containers are deployed in your local development PC (local Docker host), each microservice’s container has always an internal port (usually port 80) specified in its dockerfile, as in the following dockerfile.
+
+```csharp
```

Comment:
using csharp as the identifier for this and the next code block is breaking the formatting